### PR TITLE
CLDIDE-2648: Implement save factory without image

### PR DESCRIPTION
**Pull request by akorneta**
In case when we create a factory from the workspace inside ide we cannot choose factory image without ugly workarounds, it was proposed to add a method that consumes factory dto instance without images.

### CODE REVIEW

File: platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClient.java [hunk 1]
```diff
@@ -49,4 +51,11 @@
      * @param callback callback which returns snippet of the factory or exception if occurred
      */
     void getFactoryJson(@NotNull String workspaceId, @NotNull String path, @NotNull AsyncRequestCallback<Factory> callback);
+
+    /**
+     * Save factory to storage.
+     *
+     * @see FactoryService#saveFactory(Factory)
+     */
+    Promise<Factory> saveFactory(@NotNull Factory factory);
```
Comment (vzhukovskii) [inline platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClient.java]: Might it returns null?

Comment (akorneta) [inline platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClient.java]: no

Comment (vzhukovskii) [inline platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClient.java]: So, then add @NotNull above return type

Comment (akorneta) [inline platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClient.java]: we do not have specific rules on this, so as not to disrupt the overall design of the class I will not do it, and force this point for discussion

File: platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/impl/FactoryBaseValidator.java [hunk 2]
```diff
@@ -71,7 +71,7 @@ protected void validateSource(Factory factory) throws ConflictException {
             String parameterLocationName = "project.storage.location";
 
             // check that vcs value is correct
-            if (!("git".equals(type) || "esbwso2".equals(type))) {
+            if (!("git".equals(type) || "esbwso2".equals(type) || "github".equals(type))) {
```
Comment (skabashnyuk) [inline platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/impl/FactoryBaseValidator.java]: Need to figure out what it is.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/bootstrap/FactoryWorkspaceComponent.java [hunk 3]
```diff
@@ -157,8 +158,8 @@ public void apply(PromiseError arg) throws OperationException {
      */
     private Promise<UsersWorkspaceDto> getWorkspaceToStart() {
         final WorkspaceConfigDto workspaceConfigDto = factory.getWorkspace();
-
-        switch (factory.getPolicies().getCreate()) {
+        final String create = factory.getPolicies() == null ? "perClick" : factory.getPolicies().getCreate();
```
Comment (evoevodin) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/bootstrap/FactoryWorkspaceComponent.java]: What about moving `"perClick"` to constant like `DEFAULT_POLICY`, + i think you should rename local variable to `policy` as well, as create looks a bit weird for me =)

File: platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClientImpl.java [hunk 4]
```diff
@@ -63,4 +87,62 @@ public void getFactoryJson(String workspaceId, String path, AsyncRequestCallback
                            .header(HTTPHeader.ACCEPT, MimeType.APPLICATION_JSON)
                            .send(callback);
     }
+
+    @Override
+    public Promise<Factory> getFactoryJson(String workspaceId, String path) {
+        path = path == null ? "" : path;
```
Comment (evoevodin) [inline platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClientImpl.java]: What do you think about this?
```java
String url  = API_FACTORY_BASE_URL + "workspace/" + workspaceId;
if (path != null) {
    url += path;
}
```

Comment (akorneta) [inline platform-api-client-gwt/che-core-client-gwt-factory/src/main/java/org/eclipse/che/api/factory/gwt/client/FactoryServiceClientImpl.java]: ok

File: platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java [hunk 5]
```diff
@@ -209,6 +210,39 @@ public Factory saveFactory(Iterator<FileItem> formData,
...
+     *         instance of factory which would be stored
+     * @return decorated factory instance of which has been stored
+     * @throws org.eclipse.che.api.core.ApiException
+     *         {@link org.eclipse.che.api.core.BadRequestException} when factory is null
+     *         {@link org.eclipse.che.api.core.ConflictException} when problem occurs during factory storing
+     *         {@link org.eclipse.che.api.core.ServerException} when internal server error occurs
+     */
+    @RolesAllowed("user")
+    @POST
+    @Consumes({MediaType.APPLICATION_JSON})
```
Comment (evoevodin) [inline platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java]: You can remove `{}`

File: platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java [hunk 6]
```diff
@@ -209,6 +210,39 @@ public Factory saveFactory(Iterator<FileItem> formData,
...
+    @Produces({MediaType.APPLICATION_JSON})
+    public Factory saveFactory(Factory factory) throws ApiException {
+        final User user = EnvironmentContext.getCurrent().getUser();
+        if (user == null || user.getName() == null || user.getId() == null) {
```
Comment (evoevodin) [inline platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java]: User can't be `null` as long as `@RollesAllowed("user")` is present on top of this method

File: platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java [hunk 7]
```diff
@@ -209,6 +210,39 @@ public Factory saveFactory(Iterator<FileItem> formData,
...
+            throw new ServerException("Unable to identify user from context");
+        }
+        if (factory == null) {
```
Comment (evoevodin) [inline platform-api/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java]: It is always better to perform `null` checks on top of the checks list

**Pull request merged**